sdks/python/apache_beam/transforms/core.py

```diff
@@ -119,6 +120,112 @@ def aggregate_to(self, aggregator, input_value):
     self.state.counter_for(aggregator).update(input_value)
 
 
+class NewDoFn(WithTypeHints, HasDisplayData):
+  """A function object used by a transform with custom processing.
+
+  The ParDo transform is such a transform. The ParDo.apply
+  method will take an object of type DoFn and apply it to all elements of a
+  PCollection object.
+
+  In order to have concrete DoFn objects one has to subclass from DoFn and
+  define the desired behavior (start_bundle/finish_bundle and process) or wrap a
+  callable object using the CallableWrapperDoFn class.
+  """
+
+  ElementParam = 'ElementParam'
+  ContextParam = 'ContextParam'
+  SideInputParam = 'SideInputParam'
+  TimestampParam = 'TimestampParam'
+  WindowsParam = 'WindowsParam'
```

Comment:
I thought we decided as we wanted to preserve the optimization for the sliding windows case where the sideinputs are globally windowed as that was a performance improvement.

I'm a bit confused about the API here so some help would be great.
Lets say we have 5 windows in the element and then:
case 1 (global side inputs): Process is called once with all windows together
case 2 (windowed side inputs): Process is called 5 times once for each window

What should happen here if we only expose one window at a time ?